File: samples/flutter/lib/cowc_client.dart
/// COWC Flutter Client
///
/// A Dart client for integrating with COWC providers.

import 'dart:convert';
import 'dart:math';
import 'package:crypto/crypto.dart';
import 'package:http/http.dart' as http;

/// COWC API Client
class COWCClient {
  final String baseUrl;
  final String clientId;
  final String clientSecret;
  final String apiKey;

  String? _accessToken;
  DateTime? _tokenExpiry;

  COWCClient({
    required this.baseUrl,
    required this.clientId,
    required this.clientSecret,
    required this.apiKey,
  });

  /// Authenticate with the COWC provider
  Future<String> authenticate({
    List<String> scopes = const ['funds:read', 'portfolio:read', 'orders:write'],
  }) async {
    final response = await http.post(
      Uri.parse('$baseUrl/oauth/token'),
      headers: {'Content-Type': 'application/x-www-form-urlencoded'},
      body: {
        'grant_type': 'client_credentials',
        'client_id': clientId,
        'client_secret': clientSecret,
        'scope': scopes.join(' '),
      },
    );

    if (response.statusCode != 200) {
      throw COWCException.fromResponse(response);
    }

    final data = jsonDecode(response.body);
    _accessToken = data['access_token'];
    _tokenExpiry = DateTime.now().add(Duration(seconds: data['expires_in']));

    return _accessToken!;
  }

  bool get _isTokenExpired {
    if (_tokenExpiry == null) return true;
    return DateTime.now().isAfter(_tokenExpiry!.subtract(Duration(minutes: 5)));
  }

  Future<void> _ensureAuthenticated() async {
    if (_isTokenExpired) {
      await authenticate();
    }
  }

  Future<Map<String, dynamic>> _request(
    String method,
    String endpoint, {
    Map<String, String>? queryParams,
    Map<String, dynamic>? body,
  }) async {
    await _ensureAuthenticated();

    var uri = Uri.parse('$baseUrl$endpoint');
    if (queryParams != null) {
      uri = uri.replace(queryParameters: queryParams);
    }

    final request = http.Request(method, uri);
    request.headers.addAll({
      'Authorization': 'Bearer $_accessToken',
      'X-COWC-Partner-Key': apiKey,
      'X-COWC-Request-ID': _generateRequestId(),
      'Content-Type': 'application/json',
    });

    if (body != null) {
      request.body = jsonEncode(body);
    }

    final streamedResponse = await request.send();
    final response = await http.Response.fromStream(streamedResponse);

    final data = jsonDecode(response.body);

    if (response.statusCode >= 400) {
      throw COWCException.fromResponse(response);
    }

    return data;
  }

  String _generateRequestId() {
    final random = Random.secure();
    final values = List<int>.generate(16, (i) => random.nextInt(256));
    return values.map((b) => b.toRadixString(16).padLeft(2, '0')).join();
  }

  // Fund Discovery API

  /// List available funds
  Future<Map<String, dynamic>> listFunds({
    String? category,
    String? riskLevel,
    bool? isShariah,
    String? currency,
    int? limit,
    int? offset,
  }) async {
    final params = <String, String>{};
    if (category != null) params['category'] = category;
    if (riskLevel != null) params['risk_level'] = riskLevel;
    if (isShariah != null) params['is_shariah'] = isShariah.toString();
    if (currency != null) params['currency'] = currency;
    if (limit != null) params['limit'] = limit.toString();
    if (offset != null) params['offset'] = offset.toString();

    return _request('GET', '/funds', queryParams: params.isNotEmpty ? params : null);
  }

  /// Get fund details
  Future<Map<String, dynamic>> getFund(String fundId) async {
    return _request('GET', '/funds/$fundId');
  }

  // Portfolio API

  /// Get portfolio summary
  Future<Map<String, dynamic>> getPortfolio(String bindingId) async {
    return _request('GET', '/bindings/$bindingId/portfolio');
  }

  /// Get holdings
  Future<Map<String, dynamic>> getHoldings(String bindingId) async {
    return _request('GET', '/bindings/$bindingId/holdings');
  }

  /// Get binding status
  Future<Map<String, dynamic>> getBindingStatus(String bindingId) async {
    return _request('GET', '/bindings/$bindingId');
  }

  // Flow API

  /// Initiate onboarding flow
  Future<Map<String, dynamic>> initiateOnboarding({
    required String partnerUserId,
    String? email,
    String? phone,
    required String callbackUrl,
    String locale = 'en',
  }) async {
    return _request('POST', '/flows/onboarding', body: {
      'partner_user_id': partnerUserId,
      if (email != null) 'email': email,
      if (phone != null) 'phone': phone,
      'callback_url': callbackUrl,
      'locale': locale,
    });
  }

  /// Initiate link flow
  Future<Map<String, dynamic>> initiateLink({
    required String partnerUserId,
    required String callbackUrl,
  }) async {
    return _request('POST', '/flows/link', body: {
      'partner_user_id': partnerUserId,
      'callback_url': callbackUrl,
    });
  }

  // Order API

  /// Preview order
  Future<Map<String, dynamic>> previewOrder({
    required String bindingId,
    required String fundId,
    required double amount,
    String orderType = 'subscription',
    String currency = 'MYR',
  }) async {
    return _request('POST', '/orders/preview', body: {
      'binding_id': bindingId,
      'fund_id': fundId,
      'order_type': orderType,
      'amount': amount,
      'currency': currency,
    });
  }

  /// Create order
  Future<Map<String, dynamic>> createOrder({
    required String bindingId,
    required String fundId,
    required double amount,
    String orderType = 'subscription',
    String currency = 'MYR',
    String? partnerReferenceId,
  }) async {
    return _request('POST', '/orders', body: {
      'binding_id': bindingId,
      'fund_id': fundId,
      'order_type': orderType,
      'amount': amount,
      'currency': currency,
      if (partnerReferenceId != null) 'partner_reference_id': partnerReferenceId,
    });
  }

  /// Get order status
  Future<Map<String, dynamic>> getOrder(String orderId) async {
    return _request('GET', '/orders/$orderId');
  }
}

/// COWC API Exception
class COWCException implements Exception {
  final String message;
  final String? code;
  final Map<String, dynamic>? details;
  final int statusCode;

  COWCException({
    required this.message,
    this.code,
    this.details,
    required this.statusCode,
  });

  factory COWCException.fromResponse(http.Response response) {
    try {
      final data = jsonDecode(response.body);
      final error = data['error'] ?? {};
      return COWCException(
        message: error['message'] ?? 'Request failed',
        code: error['code'],
        details: error['details'],
        statusCode: response.statusCode,
      );
    } catch (_) {
      return COWCException(
        message: 'Request failed: ${response.statusCode}',
        statusCode: response.statusCode,
      );
    }
  }

  @override
  String toString() => 'COWCException: $message (code: $code)';
}

/// Verify webhook signature
bool verifyWebhook({
  required String payload,
  required String signature,
  required String timestamp,
  required String secret,
  int maxAge = 300,
}) {
  // Check timestamp
  final ts = int.tryParse(timestamp);
  if (ts == null) return false;

  final now = DateTime.now().millisecondsSinceEpoch ~/ 1000;
  if ((now - ts).abs() > maxAge) return false;

  // Calculate expected signature
  final signedPayload = '$timestamp.$payload';
  final hmacSha256 = Hmac(sha256, utf8.encode(secret));
  final digest = hmacSha256.convert(utf8.encode(signedPayload));
  final expected = 'sha256=${digest.toString()}';

  return signature == expected;
}

File: samples/python/cowc_client.py
"""
COWC Python Client

A Python client for integrating with COWC providers.
"""

import os
import time
import uuid
import hmac
import hashlib
from typing import Optional, Dict, Any, List
from urllib.parse import urlencode

import requests
from dotenv import load_dotenv

load_dotenv()


class COWCClient:
    """COWC API Client"""

    def __init__(
        self,
        base_url: str,
        client_id: str,
        client_secret: str,
        api_key: str
    ):
        self.base_url = base_url.rstrip('/')
        self.client_id = client_id
        self.client_secret = client_secret
        self.api_key = api_key
        self.access_token: Optional[str] = None
        self.token_expiry: Optional[float] = None

    def authenticate(self, scopes: List[str] = None) -> str:
        """
        Authenticate with the COWC provider using OAuth 2.0 Client Credentials.

        Args:
            scopes: List of requested scopes

        Returns:
            Access token
        """
        if scopes is None:
            scopes = ['funds:read', 'portfolio:read', 'orders:write']

        response = requests.post(
            f"{self.base_url}/oauth/token",
            data={
                'grant_type': 'client_credentials',
                'client_id': self.client_id,
                'client_secret': self.client_secret,
                'scope': ' '.join(scopes)
            },
            headers={'Content-Type': 'application/x-www-form-urlencoded'}
        )
        response.raise_for_status()

        data = response.json()
        self.access_token = data['access_token']
        self.token_expiry = time.time() + data['expires_in']

        return self.access_token

    def _is_token_expired(self) -> bool:
        """Check if token is expired (with 5 min buffer)"""
        if not self.token_expiry:
            return True
        return time.time() > (self.token_expiry - 300)

    def _ensure_authenticated(self):
        """Ensure we have a valid token"""
        if self._is_token_expired():
            self.authenticate()

    def _request(
        self,
        method: str,
        endpoint: str,
        params: Dict = None,
        json: Dict = None
    ) -> Dict[str, Any]:
        """Make an authenticated API request"""
        self._ensure_authenticated()

        url = f"{self.base_url}{endpoint}"
        headers = {
            'Authorization': f'Bearer {self.access_token}',
            'X-COWC-Partner-Key': self.api_key,
            'X-COWC-Request-ID': str(uuid.uuid4()),
            'Content-Type': 'application/json'
        }

        response = requests.request(
            method=method,
            url=url,
            headers=headers,
            params=params,
            json=json
        )

        data = response.json()

        if not response.ok:
            error = data.get('error', {})
            raise COWCError(
                message=error.get('message', 'Request failed'),
                code=error.get('code'),
                details=error.get('details'),
                status=response.status_code
            )

        return data

    # Fund Discovery API

    def list_funds(
        self,
        category: str = None,
        risk_level: str = None,
        is_shariah: bool = None,
        currency: str = None,
        limit: int = None,
        offset: int = None
    ) -> Dict[str, Any]:
        """List available funds"""
        params = {}
        if category:
            params['category'] = category
        if risk_level:
            params['risk_level'] = risk_level
        if is_shariah is not None:
            params['is_shariah'] = str(is_shariah).lower()
        if currency:
            params['currency'] = currency
        if limit:
            params['limit'] = limit
        if offset:
            params['offset'] = offset

        return self._request('GET', '/funds', params=params)

    def get_fund(self, fund_id: str) -> Dict[str, Any]:
        """Get fund details"""
        return self._request('GET', f'/funds/{fund_id}')

    def get_fund_nav(self, fund_id: str, from_date: str, to_date: str) -> Dict[str, Any]:
        """Get fund NAV history"""
        return self._request('GET', f'/funds/{fund_id}/nav', params={
            'from': from_date,
            'to': to_date
        })

    # Portfolio API

    def get_portfolio(self, binding_id: str) -> Dict[str, Any]:
        """Get portfolio summary"""
        return self._request('GET', f'/bindings/{binding_id}/portfolio')

    def get_holdings(self, binding_id: str) -> Dict[str, Any]:
        """Get holdings"""
        return self._request('GET', f'/bindings/{binding_id}/holdings')

    def get_transactions(
        self,
        binding_id: str,
        from_date: str = None,
        to_date: str = None,
        limit: int = None
    ) -> Dict[str, Any]:
        """Get transaction history"""
        params = {}
        if from_date:
            params['from'] = from_date
        if to_date:
            params['to'] = to_date
        if limit:
            params['limit'] = limit

        return self._request('GET', f'/bindings/{binding_id}/transactions', params=params)

    def get_binding_status(self, binding_id: str) -> Dict[str, Any]:
        """Get binding status"""
        return self._request('GET', f'/bindings/{binding_id}')

    # Order API

    def preview_order(
        self,
        binding_id: str,
        fund_id: str,
        amount: float,
        order_type: str = 'subscription',
        currency: str = 'MYR'
    ) -> Dict[str, Any]:
        """Preview an order"""
        return self._request('POST', '/orders/preview', json={
            'binding_id': binding_id,
            'fund_id': fund_id,
            'order_type': order_type,
            'amount': amount,
            'currency': currency
        })

    def create_order(
        self,
        binding_id: str,
        fund_id: str,
        amount: float,
        order_type: str = 'subscription',
        currency: str = 'MYR',
        partner_reference_id: str = None
    ) -> Dict[str, Any]:
        """Create an order"""
        return self._request('POST', '/orders', json={
            'binding_id': binding_id,
            'fund_id': fund_id,
            'order_type': order_type,
            'amount': amount,
            'currency': currency,
            'partner_reference_id': partner_reference_id
        })

    def get_order(self, order_id: str) -> Dict[str, Any]:
        """Get order status"""
        return self._request('GET', f'/orders/{order_id}')


class COWCError(Exception):
    """COWC API Error"""

    def __init__(self, message: str, code: str = None, details: Dict = None, status: int = None):
        super().__init__(message)
        self.code = code
        self.details = details
        self.status = status


def verify_webhook(
    payload: str,
    signature: str,
    timestamp: str,
    secret: str,
    max_age: int = 300
) -> bool:
    """
    Verify webhook signature.

    Args:
        payload: Raw request body
        signature: X-COWC-Signature header value
        timestamp: X-COWC-Timestamp header value
        secret: Webhook secret
        max_age: Maximum age in seconds (default: 300)

    Returns:
        True if signature is valid
    """
    # Check timestamp
    try:
        ts = int(timestamp)
        if abs(time.time() - ts) > max_age:
            return False
    except ValueError:
        return False

    # Calculate expected signature
    signed_payload = f"{timestamp}.{payload}"
    expected = 'sha256=' + hmac.new(
        secret.encode(),
        signed_payload.encode(),
        hashlib.sha256
    ).hexdigest()

    # Constant-time comparison
    return hmac.compare_digest(signature, expected)


# Example usage
if __name__ == '__main__':
    client = COWCClient(
        base_url=os.getenv('COWC_BASE_URL', 'https://staging.cashku.ai/cowc/v1'),
        client_id=os.getenv('COWC_CLIENT_ID', ''),
        client_secret=os.getenv('COWC_CLIENT_SECRET', ''),
        api_key=os.getenv('COWC_API_KEY', '')
    )

    try:
        # Authenticate
        print("Authenticating...")
        client.authenticate()
        print("✓ Authentication successful\n")

        # List funds
        print("Listing funds...")
        funds = client.list_funds(limit=5)
        print(f"✓ Found {funds.get('pagination', {}).get('total', 0)} funds\n")

        for fund in funds.get('funds', [])[:3]:
            print(f"  - {fund['fund_name']} (NAV: {fund['current_nav']})")

    except COWCError as e:
        print(f"Error: {e}")
        if e.code:
            print(f"Code: {e.code}")
    except Exception as e:
        print(f"Error: {e}")
